tests/fork/fork.c

```diff
@@ -0,0 +1,198 @@
+{
+    pid_t pid = fork();
+
+    if (pid < 0)
+    {
+        fprintf(stderr, "%s: fork() failed: %d\n", argv[0], pid);
+        exit(3);
+    }
+    else if (pid == 0)
+    {
+        _printf("*** inside child.. sleeping\n");
+        sleep(1);
+        _printf("Shutting down child\n");
+        exit(1);
+    }
+    else
+    {
+        int wstatus;
+        _printf("*** inside parent\n");
+        _printf("Waiting for child to shutdown\n");
+        if (waitpid(pid, &wstatus, 0) != pid)
+        {
+            fprintf(stderr, "waitpid on child pid did not return child pid.\n");
+            exit(1);
+        }
+        if (!WIFEXITED(wstatus))
+        {
+            fprintf(stderr, "waitpid WIFEXITED should be set.\n");
+            exit(1);
+        }
+        if (WEXITSTATUS(wstatus) != 1)
+        {
+            fprintf(
+                stderr,
+                "waitpid WEXITSTATUS should be 1 as child did exit(1).\n");
+            exit(1);
+        }
+        if (WIFSIGNALED(wstatus) != 0)
+        {
+            fprintf(stderr, "waitpid WIFSIGNALED was not sent a kill.\n");
+            exit(1);
+        }
+        _printf("Shutting down parent\n");
+    }
+    return 0;
+}
+
+int test_fork3(int argc, const char* argv[])
+{
+    pid_t pid = fork();
+
+    if (pid < 0)
+    {
+        fprintf(stderr, "%s: fork() failed: %d\n", argv[0], pid);
+        exit(3);
+    }
+    else if (pid == 0)
+    {
+        _printf("*** inside child.. sleeping and waiting for signal\n");
+        sleep(1);
+        _printf("Shutting down child\n");
+        exit(1);
+    }
+    else
+    {
+        int wstatus;
+        _printf("*** inside parent\n");
+        _printf("Sending signal to shut down child");
+        kill(pid, SIGKILL);
+        _printf("Waiting for child to shutdown\n");
+        if (waitpid(pid, &wstatus, 0) != pid)
+        {
+            fprintf(stderr, "waitpid on child pid did not return child pid.\n");
+            exit(1);
+        }
+        if (!WIFSIGNALED(wstatus))
+        {
+            fprintf(stderr, "waitpid WIFSIGNALED should be set.\n");
+            exit(1);
+        }
+        if (WTERMSIG(wstatus) != 9)
+        {
+            fprintf(
+                stderr,
+                "waitpid WTERMSIG should be 9 as child was killed via "
+                "SIGKILL).\n");
+            exit(1);
+        }
+        if (WIFSIGNALED(wstatus) == 0)
+        {
+            fprintf(
+                stderr,
+                "waitpid WIFSIGNALED should have told us child was killed.\n");
+            exit(1);
+        }
+        _printf("Shutting down parent\n");
+    }
+    return 0;
+}
+
+int main(int argc, const char* argv[])
+{
+    if (argc < 2)
+    {
+        fprintf(stderr, "Not enough args... which test?\n");
+        return -1;
+    }
+    if (strcmp(argv[1], "fork") == 0)
+    {
+        assert(test_fork1(argc, argv) == 0);
+        assert(test_fork2(argc, argv) == 0);
+        assert(test_fork3(argc, argv) == 0);
```

Comment:
Can we add a test case that print out the stack in the child? And ensure the write to the stack by the child is invisible to the parent, and vice versa?

Reply:
added 2 tests.
* the child overwrites a variable on the stack and it is unchanged in the parent. This proves the child and parent stacks are different.
* added a fork/exec test